assets/js/data/cart/push-changes.ts

```diff
@@ -0,0 +1,149 @@
+/**
+ * External dependencies
+ */
+import { debounce } from 'lodash';
+import { select, dispatch } from '@wordpress/data';
+import {
+	formatStoreApiErrorMessage,
+	pluckAddress,
+	pluckEmail,
+} from '@woocommerce/base-utils';
+import {
+	CartResponseBillingAddress,
+	CartResponseShippingAddress,
+} from '@woocommerce/type-defs/cart-response';
+import isShallowEqual from '@wordpress/is-shallow-equal';
+import { BillingAddressShippingAddress } from '@woocommerce/type-defs/cart';
+
+/**
+ * Internal dependencies
+ */
+import { STORE_KEY } from './constants';
+
+declare type CustomerData = {
+	billingData: CartResponseBillingAddress;
+	shippingAddress: CartResponseShippingAddress;
+} | null;
+
+/**
+ * Checks if a cart response contains an email property.
+ */
+const isCartResponseBillingAddress = (
+	address: CartResponseBillingAddress | CartResponseShippingAddress
+): address is CartResponseBillingAddress => {
+	return 'email' in address;
+};
+
+/**
+ * Does a shallow compare of important address data to determine if the cart needs updating on the server. This takes
+ * the current and previous address into account, as well as the billing email field.
+ */
+const isAddressDirty = <
+	T extends CartResponseBillingAddress | CartResponseShippingAddress
+>(
+	// An object containing all previous address information
+	previousAddress: T,
+	// An object containing all address information.
+	address: T
+): boolean => {
+	if (
+		isCartResponseBillingAddress( address ) &&
+		pluckEmail( address ) !==
+			pluckEmail( previousAddress as CartResponseBillingAddress )
+	) {
+		return true;
+	}
+
+	return (
+		!! address.country &&
+		! isShallowEqual(
+			pluckAddress( previousAddress ),
+			pluckAddress( address )
+		)
+	);
+};
+
+/**
+ * Local cache of customerData used for comparisons.
+ */
+let customerData = <CustomerData>null;
+
+/**
+ * Tracks which props have changed so the correct data gets pushed to the server.
+ */
+const dirtyProps = {
```

Comment:
maybe `const DIRTY_PROPS = {,,,}` for consistency ?

Reply:
Consistency with what? I see caps used for server data constants, but this is just an object tracking changes. It does change.

Reply:
I'm happy to leave this as-is unless I'm missing something :)

Reply:
I thought we have a all caps for constants approach, but looks like that's not the case.